Add new sample app component to demonstrate how to add/remove rows.

meisen45 opened the pull request:
In this cl, there are several changes:
1. The gitignore file is updated, so non-related file will not be included into this repo.
2. A new package (rows) is added to demonstrate how to allow user to publish/ remove rows from local app.
3. glide package has been updated to 4.0 for better API design and performance.

meisen45 committed:
Add new sample app component to demonstrate how to add/remove rows.

In this cl, there are several changes:
1. The gitignore file is updated, so non-related file will not be included into this repo.
2. A new package (rows) is added to demonstrate how to allow user to publish/ remove rows from local app.
3. glide package has been updated to 4.0 for better API design and performance.

meisen45 committed:
Using Guided fragment for channel customization.

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/settings/SettingsExampleFragment.java:
```diff
@@ -30,9 +30,10 @@
+        implements DialogPreference.TargetFragment {
 
-    private final Stack<Fragment> fragments = new Stack<Fragment>();
+    private static final Stack<Fragment> fragments = new Stack<Fragment>();
```
?

meisen45 replied:
Actually our settings example activity has a bug (the app will be crashed once this activity is started). I will fix this bug in an independent CL, some change was put in this CL accidentally, will be more careful next time.

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/CardPresenter.java:
```diff
@@ -0,0 +1,92 @@
+ *
+ * It will change background color/ card image accordingly
+ */
+public class CardPresenter extends Presenter {
```
VideoContentCardPresenter

meisen45 replied:
Good name, done!

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java:
```diff
@@ -0,0 +1,448 @@
+        }
+
+        @Override
+        protected void onPostExecute(Long channelId) {
```
startActivity here means the data layer needs to know which activity handle this, it's better to let data layer call listener in a Handler.

meisen45 replied:
It is a great idea, thank you for the detailed explanation. I never considered it before.

Also, for better readability, the LoadAddedChannels and ChannelPlaylistId was extracted as independent file. Please let me know if you think it is not a good idea.

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java:
```diff
@@ -0,0 +1,448 @@
+        /**
+         * A transaction id will be assigned to the added channel
+         */
+        private long addChannel(Context context, ChannelContents channelContents) {
```
The code should be inside data layer (ChannelContents), not in the UI code.

meisen45 replied:
yes, done!

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/PublishChannelFragment.java:
```diff
@@ -0,0 +1,448 @@
+    /**
+     * Async Task to remove channel from home screen
+     */
+    private final class RemoveChannelInMainScreen extends AsyncTask<ChannelContents, Void, Void> {
```
The code should be inside data layer (ChannelContents), not in the UI code.

meisen45 replied:
Yes, done. Will organize the files in a cleaner way next time.

meisen45 committed:
Extract data layer related stuff out side of UI logic

meisen45 reviewed (commented):
Thank you for your great suggestion, new patch set is uploaded

meisen45 committed:
adding documentation

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java:
```diff
@@ -0,0 +1,112 @@
+ *
+ * Usage: Just create the object (new LoadAddedChannels()), then the async task will be executed
+ */
+public class LoadAddedChannels extends AsyncTask<Void, Void, Void> {
```
move this class into ChannelContents class

meisen45 replied:
Done

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/LoadAddedChannels.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    private Listener mListener;
+
+    interface Listener {
```
The Listener should be a member of ChannelContents.  The caller who add/remove channels is not necessary the Listener.  For example,  both the BrowseFragment and GuidedStepFragment should add the Listener, whoever causes the data change is irrelevant.

meisen45 replied:
Yes, you are absolutely right. Thank you very much for the explanation!

meisen45 committed:
Change file structure, move LoadAddedChannels into data layer

also the gradle version is updated to make sure eveyr developer
can download the source code and get it compiled directly

lijingjiang commented:
A new patch set has been uploaded to address the file organization issue

meisen45 committed:
Code refactoring.

In this CL, there are several changes

1. only one video play activity is used to play video

2. Non-related async task has been removed to simplify the logic

3. Using unified interface to pass video content from different activity

4. adopting better naming convention for functions

5. channge relationship between different classes (non-static to nested)

lijingjiang commented:
A new patch set has been uploaded to address the bug we have discussed before

meisen45 committed:
Delegating video conversion task to VideoPlaybackActivity

lijingjiang commented:
New patch set has been uploaded to address the issue we talked about this morning
Thank you for reviewing my code!

dakegu2 commented on app/build.gradle:
```diff
@@ -1,13 +1,13 @@
         applicationId "android.support.v17.leanback.supportleanbackshowcase"
-        minSdkVersion 17
-        targetSdkVersion 'O'
+        minSdkVersion 23
```
keep minSdkVersion as 17

meisen45 replied:
done

dakegu2 commented on app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/rows/DynamicRowsFragment.java:
```diff
@@ -0,0 +1,237 @@
+                    intent.putExtra(VideoExampleActivity.TAG, clip);
+                    startActivity(intent);
+                } else {
+                    Intent intent = new Intent(getActivity(), ChannelPublishActivity.class);
```
show a toast "add to home screen not supported" if sdk version is too low

dakegu2 commented on app/build.gradle:
```diff
@@ -22,7 +22,7 @@ android {
 dependencies {
     compile fileTree(dir: 'libs', include: ['*.jar'])
-    compile "com.android.support:recyclerview-v7:+"
+    compile "com.android.support:recyclerview-v7:26.0.0-beta2"
```
no need to change this?

meisen45 replied:
done

jl45G committed:
minor bugs repaired

jl45G committed:
Minor bug fixed

Including the min sdk version and library version

lijingjiang commented:
New patch set has been uploaded to fix some issue about minsdk version and library version

also a toast will be created if the system version doesn't support "add to home screen" operation
